influxdb_store.go

```diff
@@ -145,6 +173,85 @@ func (in *InfluxDBStore) Trace(id ID) (*Trace, error) {
 	return trace, nil
 }
 
+func mustJSONFloat64(x interface{}) float64 {
+	n := x.(json.Number)
+	v, err := n.Float64()
+	if err != nil {
+		panic(err)
+	}
+	return v
+}
+
+func mustJSONInt64(x interface{}) int64 {
+	n := x.(json.Number)
+	v, err := n.Int64()
+	if err != nil {
+		panic(err)
+	}
+	return v
+}
+
+// Aggregate implements the Aggregator interface.
+func (in *InfluxDBStore) Aggregate(start, end time.Duration) ([]*AggregatedResult, error) {
+	// Find the mean (average), minimum, maximum, std. deviation, and count of
+	// all spans.
+	q := `SELECT MEAN("duration"),MIN("duration"),MAX("duration"),STDDEV("duration"),COUNT("duration") from spans`
+	q += fmt.Sprintf(
+		" WHERE time >= '%s' AND time <= '%s'",
+		time.Now().Add(start).UTC().Format(time.RFC3339Nano),
+		time.Now().Add(end).UTC().Format(time.RFC3339Nano),
+	)
+	q += ` GROUP BY "name"`
+	result, err := in.executeOneQuery(q)
+	if err != nil {
+		return nil, err
+	}
+
+	// Populate the results.
+	results := make([]*AggregatedResult, len(result.Series))
+	for i, row := range result.Series {
+		v := row.Values[0]
+		mean, min, max, stddev, count := v[1], v[2], v[3], v[4], v[5]
+		results[i] = &AggregatedResult{
+			RootSpanName: row.Tags["name"],
+			Average:      time.Duration(mustJSONFloat64(mean) * float64(time.Second)),
+			Min:          time.Duration(mustJSONFloat64(min) * float64(time.Second)),
+			Max:          time.Duration(mustJSONFloat64(max) * float64(time.Second)),
+			StdDev:       time.Duration(mustJSONFloat64(stddev) * float64(time.Second)),
+			Samples:      mustJSONInt64(count),
+		}
+	}
+	if len(result.Series) == 0 {
+		return nil, nil
+	}
+
+	n := 5
+	if n > len(result.Series) {
+		n = len(result.Series)
+	}
+
+	// Add in the N-slowest trace IDs for each span.
+	//
+	// TODO(slimsag): make N a pagination parameter instead.
+	result, err = in.executeOneQuery(fmt.Sprintf(`SELECT TOP("duration",%d),trace_id FROM spans GROUP BY "name"`, n))
+	if err != nil {
+		return nil, err
+	}
+	for i, row := range result.Series {
+		if row.Tags["name"] != results[i].RootSpanName {
+			panic("expectation violated") // never happens, just for sanity.
```

Comment:
Could we just return an error instead of panic?(as well in line 247); in order to avoid lib-users taking care about this.